Restart Geb VNC Container for each Test in a Spec to get correct recordings

JonasPammer opened the pull request:
Original, Upstream Open Issue: https://github.com/testcontainers/testcontainers-java/issues/3998

The other upstream solution that was suggested, i.e. https://github.com/testcontainers/testcontainers-java/pull/577, seemed way out of bound / too big of a change (that also relies on internals and commands, i.e. volitale to flv.py changes) to hack into here

i also tried a non-reflection attempt, i.e. to manage an own vnc and disable the container's interbound one -- https://gitkraken.dev/link/drafts/3d723754-a136-45c9-bb6f-8cad74600a88?type=patch. but decided to commit this here for comments.

i know reflection code will have a low chance. but at least this also contains a test one could add with Spock-Annotation of "Fails right now".

Edit: Closes #14331

JonasPammer committed:
test: proof of bug

test: proof of bug

JonasPammer committed:
chore: boolean toggle

JonasPammer committed:
fix using reflection

jdaugherty commented on grails-test-examples/geb/build.gradle:
```diff
@@ -70,9 +70,10 @@ dependencies {
-//}
+tasks.withType(Test).configureEach {
+    systemProperty('grails.geb.recording.mode', 'RECORD_ALL')
+    systemProperty('grails.geb.recording.restartPerTest', 'true')
```
Can we add a doLast { } and remove the recordings after a test run?  Otherwise it will continually add them every run locally.

cbmarcum replied:
@jdaugherty I think users would want the videos for review like reports and they would be removed by clean or cleanTests.

matrei replied:
I have suggested the following, that will keep the two last recordings:
https://github.com/JonasPammer/grails-core/pull/2/files

jdaugherty reviewed (commented):
Thank you @JonasPammer & Great Work Around!

My only concern is not cleaning up the recordings - can we clean those up unless some setting is set?  Otherwise, running grails tests locally will constantly add more recordings.

cbmarcum commented on grails-geb/src/testFixtures/groovy/grails/plugin/geb/GrailsGebSettings.groovy:
```diff
@@ -64,6 +65,7 @@ class GrailsGebSettings {
         recordingFormat = VncRecordingFormat.valueOf(
                 System.getProperty('grails.geb.recording.format', DEFAULT_RECORDING_FORMAT.name())
         )
+        recordingRestartPerTest = Boolean.parseBoolean(System.getProperty('grails.geb.recording.restartPerTest', false.toString()))
```
@jdaugherty @jamesfredley Since this is correcting default behavior that is unwanted.  Should this be enabled by default so current build files without the new setting start working correctly?

jdaugherty replied:
Yes, this probably should be the default if test recording is enabled.

cbmarcum commented:
@JonasPammer Thank you for working on this.

matrei committed:
test: improve clarity of test

And also, do not use an external url in the test.

matrei committed:
fix: rename property for clarity

matrei committed:
feat: default to restarting the recording container

matrei committed:
chore: cleanup

JonasPammer commented:
thanks, from https://github.com/JonasPammer/grails-core/pull/1#issuecomment-3214816141

jdaugherty commented on grails-test-examples/geb/build.gradle:
```diff
@@ -70,9 +70,10 @@ dependencies {
-//    //systemProperty('grails.geb.recording.mode', 'RECORD_ALL')
-//}
+tasks.withType(Test).configureEach {
+    systemProperty('grails.geb.recording.mode', 'RECORD_ALL')
```
We still need some type of cleanup of these recordings or they just grow and grow.

matrei committed:
fix: remove redundant system property setting

The system property
`grails.geb.recording.restartRecordingContainerPerTest`
was explicitly set to `true` for tests, but this is now
the default behavior, so the explicit setting is no longer
needed.

matrei committed:
test(geb): clean up old recording directories

Delete all but the two most recent recording directories
as part of the build to avoid unnecessary disk usage.

jamesfredley committed:
Merge branch '7.0.x' into jonas-issue-14331

# Conflicts:
#	grails-geb/src/testFixtures/groovy/grails/plugin/geb/GebRecordingTestListener.groovy
#	grails-geb/src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy

jamesfredley committed:
Restore Slf4j logging to WebDriverContainerHolder

Restore Slf4j logging lost in the merge
